[Windows support] Fix ycm_core_tests build

micbou opened the pull request:
Needed changes to successfully build ycm_core_tests and run it on Windows.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/227)
<!-- Reviewable:end -->

Valloric commented on cpp/ycm/DLLDefines.h:
```diff
@@ -0,0 +1,28 @@
+// Copyright (C) 2015  Google Inc.
```
Not Google Inc, is it? :)

micbou replied:
Oops, old times.

Valloric commented on cpp/ycm/tests/CMakeLists.txt:
```diff
@@ -77,13 +82,39 @@ add_executable( ${PROJECT_NAME}
+  set_target_properties( ${SERVER_LIB} PROPERTIES DEFINE_SYMBOL "TESTS_EXPORTS" )
+
+  # Fix gtest build on MSVC 11. See:
+  # http://stackoverflow.com/questions/8274588/c2977-stdtuple-too-many-template-arguments-msvc11
```
SO has short links for all answers and questions through "share" links (bottom of each question/answer). Maybe use that?

micbou replied:
Didn't know that. I'll change it.

Valloric commented on cpp/ycm/tests/CMakeLists.txt:
```diff
@@ -33,6 +33,11 @@ if ( USE_SYSTEM_GMOCK )
 else()
+  if ( WIN32 )
+    # Override BUILD_SHARED_LIBS option in gmock and gtest CMakeLists
+    set( BUILD_SHARED_LIBS ON CACHE BOOL "Build shared libraries (DLLs)." )
```
Why? I've used cmake before with gtest/gmock without resorting to shared libs. [See the FlightCrew source code](https://code.google.com/p/flightcrew/source/browse/#git) (which I wrote years ago).

micbou replied:
I tried static linking (long time ago) and could not make it work. I could try again.

Valloric replied:
No need to spend too much time on it, but taking a look at the Flightcrew `CMakeLists.txt` files should make things much easier. It's all CMake, C++ and gtest/gmock. 

These three files should be enough to get an idea: [top-level](https://github.com/Sigil-Ebook/flightcrew/blob/master/CMakeLists.txt), [library](https://github.com/Sigil-Ebook/flightcrew/blob/master/src/FlightCrew/CMakeLists.txt), [tests](https://github.com/Sigil-Ebook/flightcrew/blob/master/src/FlightCrew/tests/CMakeLists.txt).

micbou replied:
There is a difference with the Flightcrew project: the `ycm_core` and `ycm_client_support` targets are built as shared libraries. If we want to link statically `ycm_core_tests`, we also need the static version of those libraries.

Valloric replied:
Ha, excellent point. 

Feel free to proceed with the shared library approach.

Valloric commented on cpp/ycm/tests/CMakeLists.txt:
```diff
@@ -77,13 +82,39 @@ add_executable( ${PROJECT_NAME}
+endif()
+
 target_link_libraries( ${PROJECT_NAME}
+                       ${Boost_LIBRARIES}
```
In theory this shouldn't be necessary since the dependency is picked up from server/client libs.

micbou replied:
I am getting these errors without this:
```
BoostParts.lib(function.obj) : error LNK2005: "public: virtual unsigned int __cdecl boost::python::objects::py_function_impl_base::max_arity(void)const " (?max_arity@py_function_impl_base@objects@python@boost@@UEBAIXZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(function.obj) : error LNK2005: "public: virtual __cdecl boost::python::objects::py_function_impl_base::~py_function_impl_base(void)" (??1py_function_impl_base@objects@python@boost@@UEAA@XZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(errors.obj) : error LNK2005: "void __cdecl boost::python::throw_error_already_set(void)" (?throw_error_already_set@python@boost@@YAXXZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
BoostParts.lib(errors.obj) : error LNK2005: "public: virtual __cdecl boost::python::error_already_set::~error_already_set(void)" (??1error_already_set@python@boost@@UEAA@XZ) already defined in ycm_core.lib(ycm_core.pyd) [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
C:\Users\micbou\AppData\Local\Temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.exe : fatal error LNK1169: one or more multiply defined symbols found [c:\users\micbou\appdata\local\temp\ycm_build.ivs_js\ycm\tests\ycm_core_tests.vcxproj]
```
[Here](https://msdn.microsoft.com/en-us/library/72zdcz6f.aspx)'s the documentation of the error. I am trying to understand why it happens.

Valloric replied:
That looks like BoostParts is getting doubly-linked, which is bad. If adding that line fixes it, just go with that.

Valloric commented on cpp/ycm/DLLDefines.h:
```diff
@@ -0,0 +1,28 @@
+
+// We need to export symbols for gmock tests on Windows
+#if defined( _WIN32 ) && defined( TESTS_EXPORTS )
+  #define TESTS_EXPORT __declspec(dllexport)
```
Maybe TESTS_EXPORTS is not the best name; it's saying what the symbols are used for instead of what the macro does.

Possibly use DLL_EXPORT instead?

micbou replied:
I found out that CMake is already defining a preprocessor symbol (ycm_core_EXPORTS) for shared libraries if none are defined. I am updating the PR to use it instead of TESTS_EXPORTS.

Concerning the macro itself, I think DLL_EXPORT is too generic. It could already be defined (even if this is not the case). For example, Boost is defining BOOST_DLLEXPORT so I am proposing YCM_DLL_EXPORT.

micbou commented:
I replaced the preprocessor symbol TESTS_EXPORTS by the CMake predefined one. I also changed the TESTS_EXPORT name to YCM_DLL_EXPORT and I removed one unnecessary symbol export.

Valloric commented:
Reviewed 11 of 11 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/227)
<!-- Sent from Reviewable.io -->

Valloric commented:
Nice!

@homu r+

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/227)
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 38cd95d has been approved by `Valloric`

<!-- @homu r=Valloric 38cd95d48cddbb8288475b2cd5ef6c3ddd297d53 -->

homu commented:
:hourglass: Testing commit 38cd95d with merge b50be6c...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/81510672)
